core/shared/src/main/scala/fs2/internal/FreeC.scala

```diff
@@ -12,138 +12,204 @@ import scala.util.control.NonFatal
 
   def map[R2](f: R => R2): FreeC[F, R2] =
-    Bind(this,
-         (r: Either[Throwable, R]) =>
-           r match {
-             case Right(r) =>
-               try FreeC.Pure(f(r))
-               catch { case NonFatal(e) => FreeC.Fail(e) }
-             case Left(e) => FreeC.Fail(e)
-         })
+    Bind[F, R, R2](this, _.map(f).asFreeC[F])
 
   def handleErrorWith[R2 >: R](h: Throwable => FreeC[F, R2]): FreeC[F, R2] =
     Bind[F, R2, R2](this,
                     e =>
                       e match {
-                        case Right(a) => FreeC.Pure(a)
-                        case Left(e) =>
+                        case Result.Fail(e) =>
                           try h(e)
-                          catch { case NonFatal(e) => FreeC.Fail(e) }
+                          catch { case NonFatal(e) => FreeC.Result.Fail(e) }
+                        case other => other.covary[R2].asFreeC[F]
                     })
 
   def asHandler(e: Throwable): FreeC[F, R] = viewL.get match {
-    case Pure(_)    => Fail(e)
-    case Fail(e2)   => Fail(e)
-    case Bind(_, k) => k(Left(e))
+    case Result.Pure(_)  => Result.Fail(e)
+    case Result.Fail(e2) => Result.Fail(CompositeFailure(e2, e))
+    case Result.Interrupted(scope, err) =>
+      Result.Interrupted(scope, err.map(t => CompositeFailure(e, t)).orElse(Some(e)))
+    case Bind(_, k) => k(Result.Fail(e))
     case Eval(_)    => sys.error("impossible")
   }
 
   def viewL: ViewL[F, R] = mkViewL(this)
 
   def translate[G[_]](f: F ~> G): FreeC[G, R] = FreeC.suspend {
     viewL.get match {
-      case Pure(r) => Pure(r)
       case Bind(fx, k) =>
-        Bind(fx.translate(f), (e: Either[Throwable, Any]) => k(e).translate(f))
-      case Fail(e)  => Fail(e)
-      case Eval(fx) => sys.error("impossible")
+        Bind(fx.translate(f), (e: Result[Any]) => k(e).translate(f))
+      case Eval(fx)  => sys.error("impossible")
+      case Result(r) => r.asFreeC[G]
     }
   }
 }
 
 private[fs2] object FreeC {
-  final case class Pure[F[_], R](r: R) extends FreeC[F, R] {
-    override def translate[G[_]](f: F ~> G): FreeC[G, R] =
-      this.asInstanceOf[FreeC[G, R]]
-    override def toString: String = s"FreeC.Pure($r)"
+
+  def pure[F[_], A](a: A): FreeC[F, A] = Result.Pure(a)
+
+  def eval[F[_], A](f: F[A]): FreeC[F, A] = Eval(f)
+
+  def raiseError[F[_], A](rsn: Throwable): FreeC[F, A] = Result.Fail(rsn)
+
+  def interrupted[F[_], A](scopeId: Token, failure: Option[Throwable]): FreeC[F, A] =
+    Result.Interrupted(scopeId, failure)
+
+  sealed trait Result[+R] { self =>
+
+    def asFreeC[F[_]]: FreeC[F, R] = self.asInstanceOf[FreeC[F, R]]
+
+    def asExitCase: ExitCase[Throwable] = self match {
+      case Result.Pure(_)           => ExitCase.Completed
+      case Result.Fail(err)         => ExitCase.Error(err)
+      case Result.Interrupted(_, _) => ExitCase.Canceled
+    }
+
+    def covary[R2 >: R]: Result[R2] = self.asInstanceOf[Result[R2]]
```

Comment:
Don't need the cast here -- just define as `def covary[R2 >: R]: Result[R2] = self`